controllers/k8ssandra/vector.go

```diff
@@ -0,0 +1,71 @@
+package k8ssandra
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	k8ssandraapi "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/annotations"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/cassandra"
+	k8ssandralabels "github.com/k8ssandra/k8ssandra-operator/pkg/labels"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/result"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/telemetry"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *K8ssandraClusterReconciler) reconcileVector(
+	ctx context.Context,
+	kc *k8ssandraapi.K8ssandraCluster,
+	dcConfig *cassandra.DatacenterConfig,
+	remoteClient client.Client,
+	dcLogger logr.Logger,
+) result.ReconcileResult {
+	if kc.Spec.Cassandra.Telemetry.IsVectorEnabled() {
+		kcKey := utils.GetKey(kc)
+		namespace := utils.FirstNonEmptyString(dcConfig.Meta.Namespace, kc.Namespace)
+		configMapKey := client.ObjectKey{
+			Namespace: namespace,
+			Name:      telemetry.VectorAgentConfigMapName(kc.SanitizedName()),
+		}
+		// Create the vector toml config content
+		toml, err := telemetry.CreateCassandraVectorToml(ctx, kc.Spec.Cassandra.Telemetry, dcConfig, remoteClient, namespace)
+		if err != nil {
+			return result.Error(err)
+		}
+
+		desiredVectorConfigMap := telemetry.BuildVectorAgentConfigMap(namespace, kc.SanitizedName(), kc.Namespace, toml)
+		annotations.AddHashAnnotation(desiredVectorConfigMap)
+		k8ssandralabels.SetManagedBy(desiredVectorConfigMap, kcKey)
+
+		// Check if the vector config map already exists
+		actualVectorConfigMap := &corev1.ConfigMap{}
+
+		if err := remoteClient.Get(ctx, configMapKey, actualVectorConfigMap); err != nil {
+			if errors.IsNotFound(err) {
+				if err := remoteClient.Create(ctx, desiredVectorConfigMap); err != nil {
+					dcLogger.Error(err, "Failed to create Vector Agent ConfigMap")
+					return result.Error(err)
+				}
+			}
+		}
+
+		actualVectorConfigMap = actualVectorConfigMap.DeepCopy()
```

Comment:
1. I think there is a flaw here: if the cm doesn't exist and is created line 48, `actualVectorConfigMap` will be empty here, so the check line 57 will fail and an unnecessary update will be done line 61.
2. Also, you might want to handle some corner cases such as `apierrors.IsAlreadyExists`.
3. You also might want to handle the deletion case, that is, if vector was enabled, then disabled, the configmap certainly exists but we need to delete it.

I suggest that you take a look at the `reconcileDefaultPerNodeConfiguration` method since the situation is similar.

Reply:
> I suggest that you take a look at the reconcileDefaultPerNodeConfiguration method since the situation is similar.

Ok, I'll pick a few things you have there.
The workflow is slightly different so I can't use the exact same implementation.

Good catch on the line 57 failure. I've added a requeue after the creation of the config map so that we can get the actual cm on the next reconcile and be able to move forward.